tla/ccfraft.tla

```diff
@@ -997,12 +997,15 @@ HandleAppendEntriesRequest(i, j, m) ==
 HandleAppendEntriesResponse(i, j, m) ==
     /\ \/ /\ m.term = currentTerm[i]
           /\ m.success \* successful
-          /\ nextIndex'  = [nextIndex  EXCEPT ![i][j] = m.lastLogIndex + 1]
-          /\ matchIndex' = [matchIndex EXCEPT ![i][j] = m.lastLogIndex]
+          \* max(...) because why would we ever want to go backwards on a success response?!
+          /\ matchIndex' = [matchIndex EXCEPT ![i][j] = max(@, m.lastLogIndex)]
+          /\ nextIndex'  = [nextIndex  EXCEPT ![i][j] = max(@, m.lastLogIndex + 1)]
```

Comment:
Ongardie's original raft spec does *not* prevent `nextIndex` or `matchIndex` to go backwards here by taking the `max`:

https://github.com/ongardie/raft.tla/blob/974fff7236545912c035ff8041582864449d0ffe/raft.tla#L392-L396

Reply:
`nextIndex` is relatively unimportant. If it ever goes backwards due to processing an out-of-date ACK, that simply means the next AE will include redundant entries. I can potentially see ping-ponging cases where this impacts liveness, but not safety.

`matchIndex` I'm surprised by. I think we're really looking for monotonicity of the `commitIndex`, so I think `matchIndex` _could_ go backwards without violating log safety, **iff** the `AdvanceCommitIndex` calculation (and any other readers of `matchIndex`) included a max to ensure _they_ don't go backwards. But I don't see that in the original spec, so it looks like receiving out-of-order (potentially duplicated) `AppendEntryResponse` messages (ACKs) could decrement `matchIndex` and cause a decrement of `commitIndex`?

Reply:
Time permits I want to check if Ongardie's original spec also violates `MonotonicMatchIndexProp`.  If yes, we might want to seek clarification from the Raft community because the paper clearly postulates monotonicity?!

Reply:
The spec does state that both `commitIndex` and `matchIndex` are monotonic;

![image](https://github.com/microsoft/CCF/assets/4016369/1c000720-ae7a-4a45-b70a-ddea888730ce)

But makes no such claim about `nextIndex`. I agree with @eddyashton that while desirable for efficiency reasons, monotonicity of `nextIndex` isn't a safety concern. For liveness, it seems that it would be enough for it to always eventually increase, but not strictly monotonically.

Reply:
Below is a counterexample to `MonotonicMatchIndexProp` with Ongardie's original specification:

```tla
Action property MonotonicMatchIndexProp is violated.
The behavior up to this point is:
1: <Init line 125, col 9 to line 129, col 22 of module raft>
/\ messages = << >>
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = (r1 :> <<>> @@ r2 :> <<>>)
/\ state = (r1 :> Follower @@ r2 :> Follower)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 1 @@ r2 :> 1)
/\ votesResponded = (r1 :> {} @@ r2 :> {})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {})
/\ votedFor = (r1 :> Nil @@ r2 :> Nil)
2: <Timeout line 146, col 15 to line 154, col 60 of module raft>
/\ messages = << >>
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = (r1 :> <<>> @@ r2 :> <<>>)
/\ state = (r1 :> Candidate @@ r2 :> Follower)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 1)
/\ votesResponded = (r1 :> {} @@ r2 :> {})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {})
/\ votedFor = (r1 :> Nil @@ r2 :> Nil)
3: <Restart line 137, col 5 to line 143, col 57 of module raft>
/\ messages = << >>
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = (r1 :> <<>> @@ r2 :> <<>>)
/\ state = (r1 :> Candidate @@ r2 :> Follower)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 1)
/\ votesResponded = (r1 :> {} @@ r2 :> {})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {})
/\ votedFor = (r1 :> Nil @@ r2 :> Nil)
4: <Restart line 137, col 5 to line 143, col 57 of module raft>
/\ messages = << >>
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = (r1 :> <<>> @@ r2 :> <<>>)
/\ state = (r1 :> Follower @@ r2 :> Follower)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 1)
/\ votesResponded = (r1 :> {} @@ r2 :> {})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {})
/\ votedFor = (r1 :> Nil @@ r2 :> Nil)
5: <Restart line 137, col 5 to line 143, col 57 of module raft>
/\ messages = << >>
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = (r1 :> <<>> @@ r2 :> <<>>)
/\ state = (r1 :> Follower @@ r2 :> Follower)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 1)
/\ votesResponded = (r1 :> {} @@ r2 :> {})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {})
/\ votedFor = (r1 :> Nil @@ r2 :> Nil)
6: <Restart line 137, col 5 to line 143, col 57 of module raft>
/\ messages = << >>
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = (r1 :> <<>> @@ r2 :> <<>>)
/\ state = (r1 :> Follower @@ r2 :> Follower)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 1)
/\ votesResponded = (r1 :> {} @@ r2 :> {})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {})
/\ votedFor = (r1 :> Nil @@ r2 :> Nil)
7: <Timeout line 146, col 15 to line 154, col 60 of module raft>
/\ messages = << >>
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = (r1 :> <<>> @@ r2 :> <<>>)
/\ state = (r1 :> Follower @@ r2 :> Candidate)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {})
/\ votedFor = (r1 :> Nil @@ r2 :> Nil)
8: <RequestVote line 158, col 5 to line 166, col 67 of module raft>
/\ messages = ( [ mtype |-> RequestVoteRequest,
    mterm |-> 2,
    mlastLogTerm |-> 0,
    mlastLogIndex |-> 0,
    msource |-> r2,
    mdest |-> r2 ] :>
      1 )
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = (r1 :> <<>> @@ r2 :> <<>>)
/\ state = (r1 :> Follower @@ r2 :> Candidate)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {})
/\ votedFor = (r1 :> Nil @@ r2 :> Nil)
9: <Restart line 137, col 5 to line 143, col 57 of module raft>
/\ messages = ( [ mtype |-> RequestVoteRequest,
    mterm |-> 2,
    mlastLogTerm |-> 0,
    mlastLogIndex |-> 0,
    msource |-> r2,
    mdest |-> r2 ] :>
      1 )
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = (r1 :> <<>> @@ r2 :> <<>>)
/\ state = (r1 :> Follower @@ r2 :> Candidate)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {})
/\ votedFor = (r1 :> Nil @@ r2 :> Nil)
10: <Restart line 137, col 5 to line 143, col 57 of module raft>
/\ messages = ( [ mtype |-> RequestVoteRequest,
    mterm |-> 2,
    mlastLogTerm |-> 0,
    mlastLogIndex |-> 0,
    msource |-> r2,
    mdest |-> r2 ] :>
      1 )
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = (r1 :> <<>> @@ r2 :> <<>>)
/\ state = (r1 :> Follower @@ r2 :> Candidate)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {})
/\ votedFor = (r1 :> Nil @@ r2 :> Nil)
11: <RequestVote line 158, col 5 to line 166, col 67 of module raft>
/\ messages = ( [ mtype |-> RequestVoteRequest,
    mterm |-> 2,
    mlastLogTerm |-> 0,
    mlastLogIndex |-> 0,
    msource |-> r2,
    mdest |-> r1 ] :>
      1 @@
  [ mtype |-> RequestVoteRequest,
    mterm |-> 2,
    mlastLogTerm |-> 0,
    mlastLogIndex |-> 0,
    msource |-> r2,
    mdest |-> r2 ] :>
      1 )
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = (r1 :> <<>> @@ r2 :> <<>>)
/\ state = (r1 :> Follower @@ r2 :> Candidate)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {})
/\ votedFor = (r1 :> Nil @@ r2 :> Nil)
12: <Next line 428, col 12 to line 428, col 48 of module raft>
/\ messages = ( [ mtype |-> RequestVoteRequest,
    mterm |-> 2,
    mlastLogTerm |-> 0,
    mlastLogIndex |-> 0,
    msource |-> r2,
    mdest |-> r2 ] :>
      1 @@
  [ mtype |-> RequestVoteResponse,
    mterm |-> 2,
    msource |-> r1,
    mdest |-> r2,
    mvoteGranted |-> TRUE,
    mlog |-> <<>> ] :>
      1 )
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = (r1 :> <<>> @@ r2 :> <<>>)
/\ state = (r1 :> Follower @@ r2 :> Candidate)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {})
/\ votedFor = (r1 :> r2 @@ r2 :> Nil)
13: <Restart line 137, col 5 to line 143, col 57 of module raft>
/\ messages = ( [ mtype |-> RequestVoteRequest,
    mterm |-> 2,
    mlastLogTerm |-> 0,
    mlastLogIndex |-> 0,
    msource |-> r2,
    mdest |-> r2 ] :>
      1 @@
  [ mtype |-> RequestVoteResponse,
    mterm |-> 2,
    msource |-> r1,
    mdest |-> r2,
    mvoteGranted |-> TRUE,
    mlog |-> <<>> ] :>
      1 )
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = (r1 :> <<>> @@ r2 :> <<>>)
/\ state = (r1 :> Follower @@ r2 :> Candidate)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {})
/\ votedFor = (r1 :> r2 @@ r2 :> Nil)
14: <Next line 428, col 12 to line 428, col 48 of module raft>
/\ messages = ( [ mtype |-> RequestVoteRequest,
    mterm |-> 2,
    mlastLogTerm |-> 0,
    mlastLogIndex |-> 0,
    msource |-> r2,
    mdest |-> r2 ] :>
      1 )
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = (r1 :> <<>> @@ r2 :> <<>>)
/\ state = (r1 :> Follower @@ r2 :> Candidate)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {r1})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {r1})
/\ votedFor = (r1 :> r2 @@ r2 :> Nil)
15: <Restart line 137, col 5 to line 143, col 57 of module raft>
/\ messages = ( [ mtype |-> RequestVoteRequest,
    mterm |-> 2,
    mlastLogTerm |-> 0,
    mlastLogIndex |-> 0,
    msource |-> r2,
    mdest |-> r2 ] :>
      1 )
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = (r1 :> <<>> @@ r2 :> <<>>)
/\ state = (r1 :> Follower @@ r2 :> Candidate)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {r1})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {r1})
/\ votedFor = (r1 :> r2 @@ r2 :> Nil)
16: <Next line 428, col 12 to line 428, col 48 of module raft>
/\ messages = ( [ mtype |-> RequestVoteResponse,
    mterm |-> 2,
    msource |-> r2,
    mdest |-> r2,
    mvoteGranted |-> TRUE,
    mlog |-> <<>> ] :>
      1 )
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = (r1 :> <<>> @@ r2 :> <<>>)
/\ state = (r1 :> Follower @@ r2 :> Candidate)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {r1})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {r1})
/\ votedFor = (r1 :> r2 @@ r2 :> r2)
17: <Next line 428, col 12 to line 428, col 48 of module raft>
/\ messages = << >>
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = (r1 :> <<>> @@ r2 :> <<>>)
/\ state = (r1 :> Follower @@ r2 :> Candidate)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {r1, r2})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {r1, r2})
/\ votedFor = (r1 :> r2 @@ r2 :> r2)
18: <BecomeLeader line 196, col 5 to line 203, col 76 of module raft>
/\ messages = << >>
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = (r1 :> <<>> @@ r2 :> <<>>)
/\ state = (r1 :> Follower @@ r2 :> Leader)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {r1, r2})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {r1, r2})
/\ votedFor = (r1 :> r2 @@ r2 :> r2)
19: <ClientRequest line 207, col 5 to line 213, col 44 of module raft>
/\ messages = << >>
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = (r1 :> <<>> @@ r2 :> <<[term |-> 2, value |-> v2]>>)
/\ state = (r1 :> Follower @@ r2 :> Leader)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {r1, r2})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {r1, r2})
/\ votedFor = (r1 :> r2 @@ r2 :> r2)
20: <ClientRequest line 207, col 5 to line 213, col 44 of module raft>
/\ messages = << >>
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = (r1 :> <<>> @@ r2 :> <<[term |-> 2, value |-> v2], [term |-> 2, value |-> v1]>>)
/\ state = (r1 :> Follower @@ r2 :> Leader)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {r1, r2})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {r1, r2})
/\ votedFor = (r1 :> r2 @@ r2 :> r2)
21: <ClientRequest line 207, col 5 to line 213, col 44 of module raft>
/\ messages = << >>
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = ( r1 :> <<>> @@
  r2 :>
      << [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1],
         [term |-> 2, value |-> v2] >> )
/\ state = (r1 :> Follower @@ r2 :> Leader)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {r1, r2})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {r1, r2})
/\ votedFor = (r1 :> r2 @@ r2 :> r2)
22: <AppendEntries line 172, col 5 to line 192, col 67 of module raft>
/\ messages = ( [ mtype |-> AppendEntriesRequest,
    mterm |-> 2,
    msource |-> r2,
    mdest |-> r1,
    mprevLogIndex |-> 0,
    mprevLogTerm |-> 0,
    mentries |-> <<[term |-> 2, value |-> v2]>>,
    mcommitIndex |-> 0 ] :>
      1 )
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = ( r1 :> <<>> @@
  r2 :>
      << [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1],
         [term |-> 2, value |-> v2] >> )
/\ state = (r1 :> Follower @@ r2 :> Leader)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {r1, r2})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {r1, r2})
/\ votedFor = (r1 :> r2 @@ r2 :> r2)
23: <ClientRequest line 207, col 5 to line 213, col 44 of module raft>
/\ messages = ( [ mtype |-> AppendEntriesRequest,
    mterm |-> 2,
    msource |-> r2,
    mdest |-> r1,
    mprevLogIndex |-> 0,
    mprevLogTerm |-> 0,
    mentries |-> <<[term |-> 2, value |-> v2]>>,
    mcommitIndex |-> 0 ] :>
      1 )
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = ( r1 :> <<>> @@
  r2 :>
      << [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1],
         [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1] >> )
/\ state = (r1 :> Follower @@ r2 :> Leader)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {r1, r2})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {r1, r2})
/\ votedFor = (r1 :> r2 @@ r2 :> r2)
24: <AdvanceCommitIndex line 220, col 5 to line 236, col 73 of module raft>
/\ messages = ( [ mtype |-> AppendEntriesRequest,
    mterm |-> 2,
    msource |-> r2,
    mdest |-> r1,
    mprevLogIndex |-> 0,
    mprevLogTerm |-> 0,
    mentries |-> <<[term |-> 2, value |-> v2]>>,
    mcommitIndex |-> 0 ] :>
      1 )
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = ( r1 :> <<>> @@
  r2 :>
      << [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1],
         [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1] >> )
/\ state = (r1 :> Follower @@ r2 :> Leader)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {r1, r2})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {r1, r2})
/\ votedFor = (r1 :> r2 @@ r2 :> r2)
25: <Next line 428, col 12 to line 428, col 48 of module raft>
/\ messages = ( [ mtype |-> AppendEntriesRequest,
    mterm |-> 2,
    msource |-> r2,
    mdest |-> r1,
    mprevLogIndex |-> 0,
    mprevLogTerm |-> 0,
    mentries |-> <<[term |-> 2, value |-> v2]>>,
    mcommitIndex |-> 0 ] :>
      1 )
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = ( r1 :> <<[term |-> 2, value |-> v2]>> @@
  r2 :>
      << [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1],
         [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1] >> )
/\ state = (r1 :> Follower @@ r2 :> Leader)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {r1, r2})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {r1, r2})
/\ votedFor = (r1 :> r2 @@ r2 :> r2)
26: <Restart line 137, col 5 to line 143, col 57 of module raft>
/\ messages = ( [ mtype |-> AppendEntriesRequest,
    mterm |-> 2,
    msource |-> r2,
    mdest |-> r1,
    mprevLogIndex |-> 0,
    mprevLogTerm |-> 0,
    mentries |-> <<[term |-> 2, value |-> v2]>>,
    mcommitIndex |-> 0 ] :>
      1 )
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = ( r1 :> <<[term |-> 2, value |-> v2]>> @@
  r2 :>
      << [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1],
         [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1] >> )
/\ state = (r1 :> Follower @@ r2 :> Leader)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {r1, r2})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {r1, r2})
/\ votedFor = (r1 :> r2 @@ r2 :> r2)
27: <Restart line 137, col 5 to line 143, col 57 of module raft>
/\ messages = ( [ mtype |-> AppendEntriesRequest,
    mterm |-> 2,
    msource |-> r2,
    mdest |-> r1,
    mprevLogIndex |-> 0,
    mprevLogTerm |-> 0,
    mentries |-> <<[term |-> 2, value |-> v2]>>,
    mcommitIndex |-> 0 ] :>
      1 )
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = ( r1 :> <<[term |-> 2, value |-> v2]>> @@
  r2 :>
      << [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1],
         [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1] >> )
/\ state = (r1 :> Follower @@ r2 :> Leader)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {r1, r2})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {r1, r2})
/\ votedFor = (r1 :> r2 @@ r2 :> r2)
28: <ClientRequest line 207, col 5 to line 213, col 44 of module raft>
/\ messages = ( [ mtype |-> AppendEntriesRequest,
    mterm |-> 2,
    msource |-> r2,
    mdest |-> r1,
    mprevLogIndex |-> 0,
    mprevLogTerm |-> 0,
    mentries |-> <<[term |-> 2, value |-> v2]>>,
    mcommitIndex |-> 0 ] :>
      1 )
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = ( r1 :> <<[term |-> 2, value |-> v2]>> @@
  r2 :>
      << [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1],
         [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1],
         [term |-> 2, value |-> v2] >> )
/\ state = (r1 :> Follower @@ r2 :> Leader)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {r1, r2})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {r1, r2})
/\ votedFor = (r1 :> r2 @@ r2 :> r2)
29: <Restart line 137, col 5 to line 143, col 57 of module raft>
/\ messages = ( [ mtype |-> AppendEntriesRequest,
    mterm |-> 2,
    msource |-> r2,
    mdest |-> r1,
    mprevLogIndex |-> 0,
    mprevLogTerm |-> 0,
    mentries |-> <<[term |-> 2, value |-> v2]>>,
    mcommitIndex |-> 0 ] :>
      1 )
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = ( r1 :> <<[term |-> 2, value |-> v2]>> @@
  r2 :>
      << [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1],
         [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1],
         [term |-> 2, value |-> v2] >> )
/\ state = (r1 :> Follower @@ r2 :> Leader)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {r1, r2})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {r1, r2})
/\ votedFor = (r1 :> r2 @@ r2 :> r2)
30: <Restart line 137, col 5 to line 143, col 57 of module raft>
/\ messages = ( [ mtype |-> AppendEntriesRequest,
    mterm |-> 2,
    msource |-> r2,
    mdest |-> r1,
    mprevLogIndex |-> 0,
    mprevLogTerm |-> 0,
    mentries |-> <<[term |-> 2, value |-> v2]>>,
    mcommitIndex |-> 0 ] :>
      1 )
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = ( r1 :> <<[term |-> 2, value |-> v2]>> @@
  r2 :>
      << [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1],
         [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1],
         [term |-> 2, value |-> v2] >> )
/\ state = (r1 :> Follower @@ r2 :> Leader)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {r1, r2})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {r1, r2})
/\ votedFor = (r1 :> r2 @@ r2 :> r2)
31: <AppendEntries line 172, col 5 to line 192, col 67 of module raft>
/\ messages = ( [ mtype |-> AppendEntriesRequest,
    mterm |-> 2,
    msource |-> r2,
    mdest |-> r1,
    mprevLogIndex |-> 0,
    mprevLogTerm |-> 0,
    mentries |-> <<[term |-> 2, value |-> v2]>>,
    mcommitIndex |-> 0 ] :>
      2 )
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = ( r1 :> <<[term |-> 2, value |-> v2]>> @@
  r2 :>
      << [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1],
         [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1],
         [term |-> 2, value |-> v2] >> )
/\ state = (r1 :> Follower @@ r2 :> Leader)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {r1, r2})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {r1, r2})
/\ votedFor = (r1 :> r2 @@ r2 :> r2)
32: <Next line 428, col 12 to line 428, col 48 of module raft>
/\ messages = ( [ mtype |-> AppendEntriesResponse,
    mterm |-> 2,
    msource |-> r1,
    mdest |-> r2,
    msuccess |-> TRUE,
    mmatchIndex |-> 1 ] :>
      1 @@
  [ mtype |-> AppendEntriesRequest,
    mterm |-> 2,
    msource |-> r2,
    mdest |-> r1,
    mprevLogIndex |-> 0,
    mprevLogTerm |-> 0,
    mentries |-> <<[term |-> 2, value |-> v2]>>,
    mcommitIndex |-> 0 ] :>
      1 )
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = ( r1 :> <<[term |-> 2, value |-> v2]>> @@
  r2 :>
      << [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1],
         [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1],
         [term |-> 2, value |-> v2] >> )
/\ state = (r1 :> Follower @@ r2 :> Leader)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {r1, r2})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {r1, r2})
/\ votedFor = (r1 :> r2 @@ r2 :> r2)
33: <ClientRequest line 207, col 5 to line 213, col 44 of module raft>
/\ messages = ( [ mtype |-> AppendEntriesResponse,
    mterm |-> 2,
    msource |-> r1,
    mdest |-> r2,
    msuccess |-> TRUE,
    mmatchIndex |-> 1 ] :>
      1 @@
  [ mtype |-> AppendEntriesRequest,
    mterm |-> 2,
    msource |-> r2,
    mdest |-> r1,
    mprevLogIndex |-> 0,
    mprevLogTerm |-> 0,
    mentries |-> <<[term |-> 2, value |-> v2]>>,
    mcommitIndex |-> 0 ] :>
      1 )
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = ( r1 :> <<[term |-> 2, value |-> v2]>> @@
  r2 :>
      << [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1],
         [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1],
         [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v2] >> )
/\ state = (r1 :> Follower @@ r2 :> Leader)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {r1, r2})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {r1, r2})
/\ votedFor = (r1 :> r2 @@ r2 :> r2)
34: <Restart line 137, col 5 to line 143, col 57 of module raft>
/\ messages = ( [ mtype |-> AppendEntriesResponse,
    mterm |-> 2,
    msource |-> r1,
    mdest |-> r2,
    msuccess |-> TRUE,
    mmatchIndex |-> 1 ] :>
      1 @@
  [ mtype |-> AppendEntriesRequest,
    mterm |-> 2,
    msource |-> r2,
    mdest |-> r1,
    mprevLogIndex |-> 0,
    mprevLogTerm |-> 0,
    mentries |-> <<[term |-> 2, value |-> v2]>>,
    mcommitIndex |-> 0 ] :>
      1 )
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = ( r1 :> <<[term |-> 2, value |-> v2]>> @@
  r2 :>
      << [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1],
         [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1],
         [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v2] >> )
/\ state = (r1 :> Follower @@ r2 :> Leader)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {r1, r2})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {r1, r2})
/\ votedFor = (r1 :> r2 @@ r2 :> r2)
35: <Next line 428, col 12 to line 428, col 48 of module raft>
/\ messages = ( [ mtype |-> AppendEntriesResponse,
    mterm |-> 2,
    msource |-> r1,
    mdest |-> r2,
    msuccess |-> TRUE,
    mmatchIndex |-> 1 ] :>
      2 )
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = ( r1 :> <<[term |-> 2, value |-> v2]>> @@
  r2 :>
      << [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1],
         [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1],
         [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v2] >> )
/\ state = (r1 :> Follower @@ r2 :> Leader)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {r1, r2})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {r1, r2})
/\ votedFor = (r1 :> r2 @@ r2 :> r2)
36: <Next line 428, col 12 to line 428, col 48 of module raft>
/\ messages = ( [ mtype |-> AppendEntriesResponse,
    mterm |-> 2,
    msource |-> r1,
    mdest |-> r2,
    msuccess |-> TRUE,
    mmatchIndex |-> 1 ] :>
      1 )
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 1 @@ r2 :> 0))
/\ log = ( r1 :> <<[term |-> 2, value |-> v2]>> @@
  r2 :>
      << [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1],
         [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1],
         [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v2] >> )
/\ state = (r1 :> Follower @@ r2 :> Leader)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {r1, r2})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 2 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {r1, r2})
/\ votedFor = (r1 :> r2 @@ r2 :> r2)
37: <ClientRequest line 207, col 5 to line 213, col 44 of module raft>
/\ messages = ( [ mtype |-> AppendEntriesResponse,
    mterm |-> 2,
    msource |-> r1,
    mdest |-> r2,
    msuccess |-> TRUE,
    mmatchIndex |-> 1 ] :>
      1 )
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 1 @@ r2 :> 0))
/\ log = ( r1 :> <<[term |-> 2, value |-> v2]>> @@
  r2 :>
      << [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1],
         [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1],
         [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v2] >> )
/\ state = (r1 :> Follower @@ r2 :> Leader)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {r1, r2})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 2 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {r1, r2})
/\ votedFor = (r1 :> r2 @@ r2 :> r2)
38: <AppendEntries line 172, col 5 to line 192, col 67 of module raft>
/\ messages = ( [ mtype |-> AppendEntriesResponse,
    mterm |-> 2,
    msource |-> r1,
    mdest |-> r2,
    msuccess |-> TRUE,
    mmatchIndex |-> 1 ] :>
      1 @@
  [ mtype |-> AppendEntriesRequest,
    mterm |-> 2,
    msource |-> r2,
    mdest |-> r1,
    mprevLogIndex |-> 1,
    mprevLogTerm |-> 2,
    mentries |-> <<[term |-> 2, value |-> v1]>>,
    mcommitIndex |-> 0 ] :>
      1 )
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 1 @@ r2 :> 0))
/\ log = ( r1 :> <<[term |-> 2, value |-> v2]>> @@
  r2 :>
      << [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1],
         [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1],
         [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v2] >> )
/\ state = (r1 :> Follower @@ r2 :> Leader)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {r1, r2})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 2 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {r1, r2})
/\ votedFor = (r1 :> r2 @@ r2 :> r2)
39: <Restart line 137, col 5 to line 143, col 57 of module raft>
/\ messages = ( [ mtype |-> AppendEntriesResponse,
    mterm |-> 2,
    msource |-> r1,
    mdest |-> r2,
    msuccess |-> TRUE,
    mmatchIndex |-> 1 ] :>
      1 @@
  [ mtype |-> AppendEntriesRequest,
    mterm |-> 2,
    msource |-> r2,
    mdest |-> r1,
    mprevLogIndex |-> 1,
    mprevLogTerm |-> 2,
    mentries |-> <<[term |-> 2, value |-> v1]>>,
    mcommitIndex |-> 0 ] :>
      1 )
/\ matchIndex = (r1 :> (r1 :> 0 @@ r2 :> 0) @@ r2 :> (r1 :> 0 @@ r2 :> 0))
/\ log = ( r1 :> <<[term |-> 2, value |-> v2]>> @@
  r2 :>
      << [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1],
         [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v1],
         [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v2],
         [term |-> 2, value |-> v2] >> )
/\ state = (r1 :> Follower @@ r2 :> Follower)
/\ commitIndex = (r1 :> 0 @@ r2 :> 0)
/\ currentTerm = (r1 :> 2 @@ r2 :> 2)
/\ votesResponded = (r1 :> {} @@ r2 :> {})
/\ nextIndex = (r1 :> (r1 :> 1 @@ r2 :> 1) @@ r2 :> (r1 :> 1 @@ r2 :> 1))
/\ votesGranted = (r1 :> {} @@ r2 :> {})
/\ votedFor = (r1 :> r2 @@ r2 :> r2)
```

Reply:
Changing `HandleAppendEntriesResponse` to let `nextIndex` go backwards on an ACK (see below) fails trace validation of `fancy_election.1` and `suffix_collision.3`.

```diff
diff --git a/tla/ccfraft.tla b/tla/ccfraft.tla
index 6ecf302e9..17bfd353d 100644
--- a/tla/ccfraft.tla
+++ b/tla/ccfraft.tla
@@ -999,7 +999,7 @@ HandleAppendEntriesResponse(i, j, m) ==
           /\ m.success \* successful
           \* max(...) because why would we ever want to go backwards on a success response?!
           /\ matchIndex' = [matchIndex EXCEPT ![i][j] = max(@, m.lastLogIndex)]
-          /\ nextIndex'  = [nextIndex  EXCEPT ![i][j] = max(@, m.lastLogIndex + 1)]
+          /\ nextIndex'  = [nextIndex  EXCEPT ![i][j] = m.lastLogIndex + 1]
        \/ /\ \lnot m.success \* not successful
           /\ LET tm == FindHighestPossibleMatch(log[i], m.lastLogIndex, m.term)
              IN nextIndex' = [nextIndex EXCEPT ![i][j] =
```

Reply:
See https://github.com/microsoft/CCF/pull/5363